bccsp/schemes/aries/smartcard.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package aries
+
+import (
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"github.com/IBM/idemix/bccsp/types"
+	math "github.com/IBM/mathlib"
+)
+
+type Smartcard struct {
+	H0, H1, H2     *math.G1
+	Uid_sk, EID    *math.Zr
+	PRF_K0, PRF_K1 cipher.Block
+	Curve          *math.Curve
+	Rng            io.Reader
+
+	ctr int16
+}
+
+func (sc *Smartcard) PRF(input []byte, K cipher.Block) *math.Zr {
+	bytes := make([]byte, sc.Curve.ScalarByteSize)
+
+	PRF := cipher.NewCBCEncrypter(K, make([]byte, K.BlockSize()))
+	PRF.CryptBlocks(bytes, input)
+	PRF.CryptBlocks(bytes[sc.PRF_K1.BlockSize():], input)
+
+	return sc.Curve.NewZrFromBytes(bytes)
+}
+
+func (sc *Smartcard) getNymEid() (*math.Zr, *math.G1) {
+	// set PRNG_input from counter
+	PRNG_input := make([]byte, 16)
+	PRNG_input[0] = byte(sc.ctr & 0xff)
+	PRNG_input[1] = byte((sc.ctr >> 8) & 0xff)
+
+	// sample random s
+	s := sc.PRF(PRNG_input, sc.PRF_K0)
+
+	// tau0 = h0^s h2^id
+	tau0 := sc.H0.Mul2(s, sc.H2, sc.EID)
+
+	return s, tau0
+}
+
+func (sc *Smartcard) NymEid() (*math.Zr, *math.G1) {
+	// increment counter
+	sc.ctr++
```

Comment:
Just to make sure: Smartcard should not be assumed to be thread-safe. Is this written anywhere?

Reply:
ah so this code is just a smartcard emulator to test the e2e without any separate hardware. How do you think we should capture that?

Reply:
actually - a clarification: the signing part of that function is just a mock - because signing actually needs to be performed by the smartcard. The verification part instead is meaningful and used - see also next comment.

Reply:
Then, please could you a comment to the `Smartcard` struct to mention that it is used as an emulator?

Reply:
done